cohttp/src/s.ml

```diff
@@ -39,6 +39,11 @@ module type IO = sig
   type conn
   (** [conn] represents the underlying network flow *)
 
+  val refill : ic -> [ `Ok | `Eof ] t
+
+  val with_input_buffer :
+    ic -> f:(string -> pos:int -> len:int -> consume:(int -> unit) -> 'a) -> 'a
+
```

Comment:
I wanted to have an api that lets me see an immutable view within the reader's buffer, and use it to feed data to the parser. The request parser itself always runs in one shot mode and doesn't hold any state within it so this should be pretty safe. We'll either have a successfully parsed result, in which case the parser will inform us how many bytes were consumed (which will be used to inform the reader about the same), or it will respond with a message saying we need more data, or if we had an error while parsing. In both these cases none of the bytes will be consumed.

For dealing with partial reads, the intention is that we will just use `refill` again to fetch more bytes and if that succeeds we'll attempt to parse the request again.

Reply:
Why is consume a function instead of a something that we return along with `'a`?

Reply:
Hmm, no particular reason. I think `'a * int` as a return type should also work here. I'll switch to that.

Reply:
This has been switched to using a `'a * int` as the return value instead of using `consume` as a function